### PR TITLE
feat(1951): Change worker and scheduler initialization

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

There were issues with periodic build processing in  queue service:
. Periodic builds stopped working.
   1. We noticed that while build schedule times are being updated, periodic build did not get created. Scheduler did not even call create event end point to create those builds.
   2. We suspect it's caused by the way the scheduler/worker is initialized.
   3. Also need to evaluate error handling, whether it stops working in response to an error
2. Circuit breaker errors for large pipeline.

## Objective

This PR address these issues and changes the way worker and scheduler are initialized.
Further adds a shutdown plugin for queue cleanup

## References

https://github.com/screwdriver-cd/screwdriver/issues/1951#issuecomment-609938807

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/shutdown.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                        logger.info(`executing task ${key}`);
+                        const item = tasks[key];
+
+                        await item.task();
+                    })
+                );
+
+                return Promise.resolve();
+            } catch (err) {
+                logger.error(err);
```
Comment (jithine) [inline plugins/shutdown.js:53]: add a message for easy searching.

File: plugins/shutdown.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        const gracefulStop = async () => {
+            try {
+                logger.info('gracefully shutting down server');
+                await server.stop({
+                    timeout: 5000
+                });
+                process.exit(0);
+            } catch (err) {
+                logger.error(err);
```
Comment (jithine) [inline plugins/shutdown.js:66]: add a message for easy searching.

File: lib/server.js [hunk 3]
```diff
@@ -93,28 +92,31 @@ module.exports = async config => {
             return server.plugins.auth.generateToken(profile);
         };
 
-        // init queue
-        const res1 = await server.inject({
```
Comment (jithine) [inline lib/server.js:97]: was this making an api POST call to itself when server starts ?

File: plugins/queue/scheduler.js [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+            if (eventId) {
+                buildEvent.parentEventId = eventId;
+            }
+            if (buildId) {
+                buildEvent.buildId = buildId;
+            }
+
+            await helper.createBuildEvent(apiUri, jwt, buildEvent, executor.requestRetryStrategyPostEvent);
+        } else {
+            logger.error(`POST event for pipeline failed as no admin found: ${pipelineId}:${job.name}:${job.id}`);
```
Comment (jithine) [inline plugins/queue/scheduler.js:63]: shouldn't this throw an error ?

File: plugins/queue/scheduler.js [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+
+    if (triggerBuild) {
+        config.causeMessage = 'Started by periodic build scheduler';
+
+        // Even if post event failed for this event after retry, we should still enqueue the next event
+        try {
+            await postBuildEvent(executor, config);
+        } catch (err) {
+            logger.error(
+                `periodic builds: failed to post build event for job ${job.id} in pipeline ${pipeline.id}: ${err}`
```
Comment (jithine) [inline plugins/queue/scheduler.js:164]: `err` can be the second argument for `logger.error`

File: plugins/queue/scheduler.js [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+
+        const token = executor.tokenGen({
+            pipelineId,
+            service: 'queue',
+            jobId: job.id,
+            scmContext: pipeline.scmContext,
+            scope: ['user']
+        });
+
+        const admin = await helper.getPipelineAdmin(token, apiUri, pipelineId, executor.requestRetryStrategy);
```
Comment (jithine) [inline plugins/queue/scheduler.js:37]: error handling for getPipelineAdmin is wrong https://github.com/screwdriver-cd/queue-service/blob/master/plugins/helper.js#L178-L180

File: plugins/queue/scheduler.js [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+    const origTime = new Date(currentTime.getTime());
+
+    timeOutOfWindows(freezeWindows, currentTime);
+
+    let enq;
+
+    // Check freeze window
+    if (currentTime.getTime() > origTime.getTime() && !forceStart) {
+        await helper
+            .updateBuild(
```
Comment (jithine) [inline plugins/queue/scheduler.js:272]: error handling is wtong https://github.com/screwdriver-cd/queue-service/blob/master/plugins/helper.js#L209-L211

File: plugins/queue/scheduler.js [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+ * @method startTimer
+ * @param  {Object} config               Configuration
+ * @param  {String} config.buildId       Unique ID for a build
+ * @param  {String} config.startTime     Start time fo build
+ * @param  {String} config.buildStatus     Status of build
+ * @return {Promise}
+ */
+async function startTimer(executor, config) {
+    try {
+        await executor.connect();
```
Comment (jithine) [inline plugins/queue/scheduler.js:478]: we are using `connect` in all places, but what's error handling for `connect` ? https://github.com/screwdriver-cd/queue-service/blob/master/lib/queue.js#L76 Looks like if not connected it merely returns and subsequent redis commands will fail. `connect` has to ensure it's connected or will re-connect, else error out.

File: lib/server.js (hunk 3 above)
Comment (pritamstyz4ever) [reply lib/server.js:97]: Yes, server.inject

File: plugins/shutdown.js [hunk 9]
```diff
@@ -57,28 +57,28 @@ const shutdownPlugin = {
...
         };
 
         const onSigterm = async () => {
             try {
-                logger.info('got SIGTERM; running triggers before shutdown');
+                logger.info('shutdown-> got SIGTERM; running triggers before shutdown');
                 const res = await promiseTimeout(taskHandler(), terminationGracePeriod * 1000);
 
                 if (res) {
                     logger.error(res);
```
Comment (jithine) [inline plugins/shutdown.js:77]: missed one to update log message

File: plugins/queue/scheduler.js (hunk 5 above)
Comment (pritamstyz4ever) [reply plugins/queue/scheduler.js:164]: ok, sometimes logger.error doesn't escape the err object and prints garbage

File: plugins/queue/scheduler.js [hunk 10]
```diff
@@ -462,6 +464,8 @@ async function init(executor) {
     await executor.multiWorker.start();
     await executor.scheduler.connect();
     await executor.scheduler.start();
+
+    return Promise.resolve('Scheduler started');
```
Comment (jithine) [inline plugins/queue/scheduler.js:468]: you can just return 'Scheduler starter` no need to wrap it in `Promise`

File: lib/queue.js [hunk 11]
```diff
@@ -69,11 +70,16 @@ module.exports = class ExecutorQueue {
...
+        try {
+            if (this.queue.connection.connected) {
+                return Promise.resolve();
+            }
 
-        return this.queueBreaker.runCommand('connect');
+            return this.queueBreaker.runCommand('connect');
+        } catch (err) {
+            logger.error('Failed to connect to redis', err);
+            throw err;
```
Comment (jithine) [inline lib/queue.js:81]: can we test this scenario ? how does queue behave on connect error.

**Pull request merged**